compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -259,44 +260,105 @@ impl<'hir> LoweringContext<'_, 'hir> {
-        let call = self.arena.alloc(self.mk_expr(hir::ExprKind::Call(path_expr, args), span));
+        let has_generic_args =
+            delegation.path.segments.iter().rev().skip(1).any(|segment| segment.args.is_some());
+
+        let (stmts, call, block_id) = if self
+            .get_resolution_id(delegation.id, span)
+            .and_then(|def_id| Ok(self.has_self(def_id, span)))
+            .unwrap_or_default()
+            && delegation.qself.is_none()
+            && !has_generic_args
+        {
+            let ast_segment = delegation.path.segments.last().unwrap();
+            let segment = self.lower_path_segment(
+                delegation.path.span,
+                ast_segment,
+                ParamMode::Optional,
+                ParenthesizedGenericArgs::Err,
+                ImplTraitContext::Disallowed(ImplTraitPosition::Path),
+                None,
+                None,
+            );
+            let segment = self.arena.alloc(segment);
+
+            let args = &*self.arena.alloc_from_iter(args);
+            let (stmts, receiver, args, block_id) = if let Some(target_expr) = target_expr {
+                let receiver = target_expr.expr.unwrap();
+                (Some(target_expr.stmts), receiver, args, target_expr.hir_id)
+            } else {
+                (None, &args[0], &args[1..], self.next_id())
+            };
+
+            let method_call_id = self.next_id();
+            if let Some(traits) = self.resolver.trait_map.remove(&delegation.id) {
+                self.trait_map.insert(method_call_id.local_id, traits.into_boxed_slice());
+            }
 
+            let method_call = self.arena.alloc(hir::Expr {
+                hir_id: method_call_id,
+                kind: hir::ExprKind::MethodCall(segment, receiver, args, span),
+                span,
+            });
+
+            (stmts, method_call, block_id)
+        } else {
+            let path = self.lower_qpath(
+                delegation.id,
+                &delegation.qself,
+                &delegation.path,
+                ParamMode::Optional,
+                ImplTraitContext::Disallowed(ImplTraitPosition::Path),
+                None,
+            );
+
+            if let Some(target_expr) = target_expr {
+                let target_expr = self.arena.alloc(target_expr);
+                let fst_arg =
+                    self.mk_expr(hir::ExprKind::Block(target_expr, None), target_expr.span);
+                args.insert(0, fst_arg);
+            }
+
+            let args = self.arena.alloc_from_iter(args);
+            let callee_path = self.arena.alloc(self.mk_expr(hir::ExprKind::Path(path), span));
+
+            let call = self.arena.alloc(self.mk_expr(hir::ExprKind::Call(callee_path, args), span));
+
+            (None, call, self.next_id())
+        };
         let block = self.arena.alloc(hir::Block {
-            stmts: &[],
+            stmts: stmts.unwrap_or(&[]),
```

Comment:
What is the desugaring for
```rust
reuse foo { stmt1; stmt2; expr }
```
?

```rust
{
  foo({ stmt1; stmt2; expr })
}
```
or
```rust
{
  stmt1;
  stmt2;
  foo(expr)
}
```

I would actually expect the first desugaring, but then the `stmt` field should always be empty.

Reply:
Also, is the "target expression unblocking" `call({ self.0 }) -> `call(self.0)`  required for the test suite to pass?

Reply:
The second approach was previously taken. For now, target expression is lowered into block except for a cases with single tail expression. See `lower_target_expr`.